### ISSUE TITLE
Questions about pytorch3d gradient behavior

### ISSUE THREAD

**Issue by aluo-x** [facebookresearch/pytorch3d#417]
I have been going over pytorch3d's rasterization (python version) & blending code to make sure my mental assumptions were correct. I have read over the pytorch3d tech report, and the softras paper, but still had a few confusions. Would appreciate any clarifications.

For zbuf:
1. the zbuf output is a hard assignment (1 face per pixel in xy direction), and if I took only the `top_k=1` of zbuf, it would only be differentiable with respect to the nearest face (within some "radius"). 
2. If I instead, colored the mesh/vertices using the distance of each vertex to the camera plane and used `softmax_rgb_blend`, the new rendered "zbuf" would then be differentiable w.r.t. nearby faces (in both xy and z directions).
3. Not sure how to phrase this, but I wonder if during optimization (using L1 or L2), I should mask the target & produced depth maps with the intersection of the valid regions, so the "background color" does not produce spurious depth gradients

Generic:
1. pytorch3d has a different face assignment logic than Softras, where pytorch3d performs soft assignment for faces only within a certain "radius", whereas Softras performs soft assignment for all faces against all pixels?
2. I understand ultimately all the gradients are approximations, does this optimization significantly affect the optimization? Or does it not matter because for far-away faces, they have small weight anyways?
3. I notice that pytorch3d uses the nearest euclidean distance (edge) for each face/pixel, would more approximate solutions work? For example instead the triangle center was used?

Difference between `sigmoid_alpha_blend` and `softmax_rgb_blend` for color output only:
1. In sigmoid, the gradients for a pixel exists for faces close in xy (but not z)? But for softmax, the gradients for a pixel exists for faces close in xy (and top k along the z dimension)
2. Or should I never use `sigmoid_alpha_blend` for color optimization? 

Difference between `sigmoid_alpha_blend` and `softmax_rgb_blend` for silhouette output:
1. Is there any meaninful difference for the alpha masks produced between these two methods? Or can they both be used for silhouette optimization?

**Comment by nikhilaravi** [comment]
@aluo-x these are all great questions! 

### Q1 
> Not sure how to phrase this, but I wonder if during optimization (using L1 or L2), I should mask the target & produced depth maps with the intersection of the valid regions, so the "background color" does not produce spurious depth gradients

What are you trying to learn? The background faces should all give zero gradients for the mesh vertices from rasterization. The gradients for the vertices are initialized to zero:
https://github.com/facebookresearch/pytorch3d/blob/3b035f57f08295efc9af076ea60f62ad26d88b91/pytorch3d/csrc/rasterize_meshes/rasterize_meshes.cu#L544

This is the step in the backward CUDA kernel where we gradients are calculated - we ignore all the background faces (i.e. faces which have `pix_to_face = -1`)

https://github.com/facebookresearch/pytorch3d/blob/3b035f57f08295efc9af076ea60f62ad26d88b91/pytorch3d/csrc/rasterize_meshes/rasterize_meshes.cu#L418-L421

### Q2
> pytorch3d has a different face assignment logic than Softras, where pytorch3d performs soft assignment for faces only within a certain "radius", whereas Softras performs soft assignment for all faces against all pixels?

Actually SoftRas also has a radius parameter called `dist_eps` which they use to expand the bounding box of a triangle. 

This is where the `dist_eps` value is initialized:
- In the PyTorch wrapper:

https://github.com/ShichenLiu/SoftRas/blob/e14cadd904c78d03851caf0cbd822173626dd82a/soft_renderer/functional/soft_rasterize.py#L35 

- In the CUDA kernel it is used to create a `threshold`:

https://github.com/ShichenLiu/SoftRas/blob/f6442793d0b3a4e02e76c2dd3a0209e1ac6c278b/soft_renderer/cuda/soft_rasterize_cuda_kernel.cu#L509

In the CUDA kernel this is where the `threshold` is used to check if the pixel lies within the bbox of the triangle: 

https://github.com/ShichenLiu/SoftRas/blob/f6442793d0b3a4e02e76c2dd3a0209e1ac6c278b/soft_renderer/cuda/soft_rasterize_cuda_kernel.cu#L33-L38

The key difference between SoftRas and PyTorch3D is that Softras performs soft assignment for all faces against all pixels whereas PyTorch3D you can limit the number of faces with the `faces_per_pixel` setting. 


### Q3 
> I notice that pytorch3d uses the nearest euclidean distance (edge) for each face/pixel, would more approximate solutions work? For example instead the triangle center was used?

This is something that we have not experimented with! Euclidean distance of the pixel from the closest triangle edge results in positive distances when the pixel is outside the face and negative if the pixel is inside the face - in the blending step this enables a "soft" boundary on the outside surface of the mesh.  If for all pixels you measure the distance from the triangle center then you won't get the same behavior with the current blending functions as all distances will be positive - you'll need to rethink how the distances are used. 


### Q4
> Or should I never use `sigmoid_alpha_blend` for color optimization?

Yes `sigmoid_alpha_blend` only blends the alpha channel so will ignore the RGB values and isn't suitable for color optimization. 


### Q5 
> Is there any meaningful difference for the alpha masks produced between these two methods? Or can they both be used for silhouette optimization?

The alpha masks should be the same for both functions - currently for `sigmoid_alpha_blend` the alpha is calculated using a CUDA kernel function but it should give exactly the same output/gradients as the PyTorch implementation in `softmax_rgb_blend`.

**Comment by aluo-x** [comment]
Thanks for the response. This really helped clarify a lot of my questions. To answer myself, I think that a mask (of the intersection between the rendered and gt 2.5D depth map) should be used when computing the loss, or else the background region could affect the gradients incorrectly.  

The cited code samples are very useful. On the topic of `threshold`, I guess I was misled by [Figure 5 of the softras](https://arxiv.org/abs/1904.01786) paper. I interpreted that as all pixels against all triangles (in the xy direction), but did not realize that a threshold is used. 

Closing this issue for now.
